### DISCUSSION TITLE
#[serde_as(as = Option<...>)] does not allow field to be missing

### DISCUSSION THREAD

**Discussion by chyvonomys** [jonasbb/serde_with#183] [Q&A] [answered]
I'm trying to deserialize one optional field from comma-separated string:
```rust
#[serde_as]
#[derive(Deserialize, Debug)]
struct Query {
    search: Option<String>,
    #[serde_as(as = "Option<StringWithSeparator::<CommaSeparator, u32>>")]
    priority: Option<Vec<u32>>,
}

#[test]
fn test_queries() {
    let query1 = "priority=2,3";
    let q1: Result<Query, _> = serde_urlencoded::from_str(query1);
    println!("ok: {:?}", q1);
    let query2 = "search=ttt";
    let q2: Result<Query, _> = serde_urlencoded::from_str(query2);
    println!("fail: {:?}", q2);
}
```

But it doesn't allow that field to be missing:

```
ok: Ok(Query { search: None, priority: Some([2, 3]) })
fail: Err(Error { err: "missing field `priority`" })
```

**Comment by jonasbb** [comment, marked as answer]
This unfortunately a side effect of using `deserialize_with`. I see how I can better document the pitfall.

If you provide serde's `default` attribute, then it behaves as expected. It will deserialize a `None` in this case:

```rust
#[serde_as]
#[derive(Deserialize, Debug)]
struct Query {
    search: Option<String>,
    #[serde_as(as = "Option<StringWithSeparator::<CommaSeparator, u32>>")]
    #[serde(default)]
    priority: Option<Vec<u32>>,
}
```

If you do not care about the distinction between `priority` was provided but the list is empty and `priority` is missing, you can also remove the `Option` from `priority`:

```rust
#[serde_as(as = "StringWithSeparator::<CommaSeparator, u32>")]
#[serde(default, skip_serializing_if = "Vec::is_empty")]
priority: Vec<u32>,
```

The `skip_serializing_if` is only necessary if you also want to use `Serialize`.

I hope this helps you :)

**Comment by chyvonomys** [comment]
Thanks! First solution works for me.
It would be nice if that was the default behavior (without extra attribute), if possible.

**Comment by jonasbb** [comment]
Thanks for the feedback, I will consider that as an option.
